text/0000-float-next-up-down.md

```diff
@@ -0,0 +1,333 @@
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Two more functions get added to `f32` and `f64`, which may be considered
+already cluttered by some.
+
+Additionally, there is a minor pitfall regarding signed zero. Repeatedly calling
+`next_up` on a negative number will iterate over all values above it, with the
+exception of +0.0, only -0.0 will be visited. Similarly starting at positive
+number and iterating downwards will only visit +0.0, not -0.0.
+
+However, if we were to define `(-0.0).next_up() == 0.0` we would lose compliance
+with the IEEE 754 standard, and lose the property that `x.next_up() > x` for all
+finite `x`. It would also lead to the pitfall that `(0.0).next_down()` would not
+be the smallest negative number, but -0.0 instead.
+
+Finally, there is a minor risk of confusion regarding precedence with unary
+minus. A user might inadvertently write `-1.0.next_up()` instead of
+`(-1.0).next_up()`, giving a value on the wrong side of -1. However, this
+potential confusion holds for most methods on `f32`/`f64`, and can be avoided
+by the cautious by writing `f32::next_up(-1.0)`.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+To implement the features described in the motivation the user essentially
+*needs* the `next_up`/`next_down` methods, or the alternative mentioned just
+below. If these are not available the user must either install a third party
+library for what is essentially one elementary function, or implement it
+themselves using `to_bits` and `from_bits`. This has several issues or pitfalls:
+
+ 1. The user might not even be aware that a third party library exists,
+    searching the standard library in vain. If they find a third party library
+    they might not be able to judge if it is of sufficient quality and with the
+    exact semantics they expect.
+
+ 2. Even if the user is aware of IEEE 754 representation and chooses to
+    implement it themselves, they might not get the edge cases correct. It is
+    also a wasted duplicate effort.
+
+ 3. The user might misunderstand the meaning of `f32::EPSILON`, thinking that
+    adding this to a number results in the next floating point number.
+    Alternatively they might misunderstand `f32::MIN_POSITIVE` to be the
+    smallest positive `f32`, or believe that `x + f32::MIN_POSITIVE` is a
+    correct implementation of `x.next_up()`.
+  
+ 4. The user might give up entirely and simply choose an arbitrary offset, e.g.
+    instead of `x.clamp(0, 1.0.next_down())` they end up writing
+    `x.clamp(0, 1.0 - 1e-9)`.
+
+The main alternative to these two functions is `nextafter(x, y)` (sometimes
+called `nexttoward`). This function was specified in IEEE 754-1985 to "return
+the next representable neighbor of `x` in the direction toward `y`". If `x == y`
+then `x` is supposed to be returned. Besides error signaling and NaNs, that is
+the complete specification.
+
+We did not choose this function for three reasons:
```

Comment:
`next_up` and `next_down` are also easier to implement more efficiently as they don't need to dispatch algorithm based on 2 inputs.